### PR TITLE
Fix for bug #1163

### LINKED ISSUE #1163: kops platform issue with k8s service selectors
**What happened?**
There is a bug in release `release-0.8` in kops platform.

**Did you expect to see some different?**
Selectors for scheduler, controllerManager and kube-dns are wrong.

**How to reproduce it (as minimally and precisely as possible)**:
Use release-0.8 to create manifests for `kube-prometheus/platforms/kops.libsonnet`.

**Environment**

* Prometheus Operator version:

    `0.47.0`
    

* Kubernetes version information:

    `1.20.7`
    

* Kubernetes cluster kind:

    `kops`

**Anything else we need to know?**:
Label key should be `k8s-app` instead of `app.kubernetes.io/name`.

### PR DESCRIPTION
**Pull request by a8j8i8t8**
This closes #1163

### CODE REVIEW

File: jsonnet/kube-prometheus/platforms/kops.libsonnet [hunk 1]
```diff
@@ -18,22 +18,22 @@ local service(name, namespace, labels, selector, ports) = {
     kubeControllerManagerPrometheusDiscoveryService: service(
       'kube-controller-manager-prometheus-discovery',
       'kube-system',
-      { 'app.kubernetes.io/name': 'kube-controller-manager' },
-      { 'app.kubernetes.io/name': 'kube-controller-manager' },
+      { 'k8s-app': 'kube-controller-manager' },
```
Comment (paulfantom) [inline jsonnet/kube-prometheus/platforms/kops.libsonnet:21]: To allow smooth operations, this line should contain both labels, `app.kubernetes.io/name` and `k8s-app` as it is passed to metadata and not to the selector.

File: jsonnet/kube-prometheus/platforms/kops.libsonnet [hunk 2]
```diff
@@ -18,22 +18,22 @@ local service(name, namespace, labels, selector, ports) = {
...
+      { 'k8s-app': 'kube-controller-manager' },
       [{ name: 'https-metrics', port: 10257, targetPort: 10257 }]
     ),
     kubeSchedulerPrometheusDiscoveryService: service(
       'kube-scheduler-prometheus-discovery',
       'kube-system',
-      { 'app.kubernetes.io/name': 'kube-scheduler' },
-      { 'app.kubernetes.io/name': 'kube-scheduler' },
+      { 'k8s-app': 'kube-scheduler' },
```
Comment (paulfantom) [inline jsonnet/kube-prometheus/platforms/kops.libsonnet:28]: Same as above

File: jsonnet/kube-prometheus/platforms/kops.libsonnet [hunk 3]
```diff
@@ -18,22 +18,22 @@ local service(name, namespace, labels, selector, ports) = {
...
+      { 'k8s-app': 'kube-scheduler' },
       [{ name: 'https-metrics', port: 10259, targetPort: 10259 }]
     ),
     kubeDnsPrometheusDiscoveryService: service(
       'kube-dns-prometheus-discovery',
       'kube-system',
-      { 'app.kubernetes.io/name': 'kube-dns' },
-      { 'app.kubernetes.io/name': 'kube-dns' },
+      { 'k8s-app': 'kube-dns' },
```
Comment (paulfantom) [inline jsonnet/kube-prometheus/platforms/kops.libsonnet:35]: Same as above

**Review by paulfantom — CHANGES_REQUESTED** [review]
Could we first make this change on `main` branch and backport? Otherwise, we will get into the same issue in few months.

File: jsonnet/kube-prometheus/platforms/kops.libsonnet (hunk 2 above)
Comment (a8j8i8t8) [reply jsonnet/kube-prometheus/platforms/kops.libsonnet:30]: @paulfantom Kept both labels and change base branch to `main`.

File: jsonnet/kube-prometheus/addons/ksm-lite.libsonnet [hunk 4]
```diff
@@ -16,8 +16,8 @@ local addArgs(args, name, containers) = std.map(
             containers: addArgs(
               [|||
                 --metric-denylist=
-                kube_*_created,
-                kube_*_metadata_resource_version,
+                kube_.+_created,
+                kube_.+_metadata_resource_version,
```
Comment (paulfantom) [inline jsonnet/kube-prometheus/addons/ksm-lite.libsonnet:20]: [Those are already on `main` branch](https://github.com/prometheus-operator/kube-prometheus/blob/main/jsonnet/kube-prometheus/addons/ksm-lite.libsonnet#L19-L20) and shouldn't be added again.

File: jsonnet/kube-prometheus/components/prometheus-adapter.libsonnet [hunk 5]
```diff
@@ -300,4 +300,20 @@ function(params) {
...
+      namespace: pa._config.namespace,
+      labels: pa._config.commonLabels,
+    },
+    spec: {
+      minAvailable: 1,
+      selector: {
+        matchLabels: pa._config.selectorLabels,
+      },
+    },
+  },
```
Comment (paulfantom) [inline jsonnet/kube-prometheus/components/prometheus-adapter.libsonnet:318]: This shouldn't be a part of this PR.

File: jsonnet/kube-prometheus/platforms/kops.libsonnet (hunk 1 above)
Comment (paulfantom) [reply jsonnet/kube-prometheus/platforms/kops.libsonnet:21]: Here we are passing parameters to a `service()` function. This function accepts 5 parameters and cannot be extended by appending new ones. This should be written as (I added comments to explain what is each parameter doing):
```jsonnet
kubeControllerManagerPrometheusDiscoveryService: service(
  'kube-controller-manager-prometheus-discovery',  # name of the Service object
  'kube-system',  # namespace in which Service object is placed
  { 'k8s-app': 'kube-controller-manager', 'app.kubernetes.io/name': 'kube-controller-manager' }  # labels added as metadata to the Service object
  { 'k8s-app': 'kube-controller-manager' },  # labels used in Service selector
  [{ name: 'https-metrics', port: 10257, targetPort: 10257 }]  # list of ports for the Service
),
```

The above example will create a Service with 2 labels that will match Pods with `k8s-app: kube-controller-manager` label set.

Pattern like this should be extended to the other Services from this addon.

File: jsonnet/kube-prometheus/platforms/kops.libsonnet (hunk 2 above)
Comment (paulfantom) [reply jsonnet/kube-prometheus/platforms/kops.libsonnet:30]: See my comment about `kube-controller-manager` as it is applicable also here.

File: jsonnet/kube-prometheus/platforms/kops.libsonnet (hunk 3 above)
Comment (paulfantom) [reply jsonnet/kube-prometheus/platforms/kops.libsonnet:39]: See my comment about `kube-controller-manager` as it is applicable also here.

**Review by paulfantom — CHANGES_REQUESTED** [review]
It seems that rebase went wrong and PR now includes commits that are already on `main` branch.

File: jsonnet/kube-prometheus/platforms/kops.libsonnet (hunk 1 above)
Comment (a8j8i8t8) [reply jsonnet/kube-prometheus/platforms/kops.libsonnet]: Oh this was stupid mistake, fixed it.

File: jsonnet/kube-prometheus/addons/ksm-lite.libsonnet (hunk 4 above)
Comment (a8j8i8t8) [reply jsonnet/kube-prometheus/addons/ksm-lite.libsonnet]: Removed this from my branch

File: jsonnet/kube-prometheus/components/prometheus-adapter.libsonnet (hunk 5 above)
Comment (a8j8i8t8) [reply jsonnet/kube-prometheus/components/prometheus-adapter.libsonnet]: Removed this from my branch

**Comment by paulfantom** [comment]
Squashed and merged to fix git history.

Thanks! :100:

**Comment by dmcnaught** [comment]
@paulfantom I see this fixed in main, can we also fix in release-0.8?

**Comment by a8j8i8t8** [comment]
If you can that'd be awesome @dmcnaught @paulfantom

**Comment by paulfantom** [comment]
@dmcnaught @a8j8i8t8 feel free to create a backport PR :)

**Comment by dmcnaught** [comment]
https://github.com/prometheus-operator/kube-prometheus/pull/1189

**Pull request merged**